Added stacklevel=2 to all warnings.warn calls (issue 10643)

hershen opened the pull request:
## PR Summary
Fix issue #10643
1) Added stacklevel=2 to all warnings.warn calls that didn't have stacklevel set. 
2) Added new section to contributing.rst advising to set stacklevel when using warnings.warn

Note:
I liked @ImportanceOfBeingErnest 's suggestion of advising to set stacklevel in the documentation. I didn't find a good place for it, so I created a new section in contributing.rst. Please let me know if there's a better place or if it's too long.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

anntzer commented:
In general, this is an improvement.  But an even better improvement would be to just use _warn_external everywhere -- see https://github.com/matplotlib/matplotlib/pull/11298 (and adapt the (nice) explanation for it).  (See for example (not in this PR) backend_pdf's writeInfoDict which currently warns with stacklevel=2 on invalid keywords, but that'll point to some other internal method of backend_pdf which only makes things more (not less) confusing for the end user.)

However, it would be even better *not* to apply either stacklevel=2, or _warn_external, indiscriminately.  For example, the change in backend_bases.py in this PR is in key_press_handler, which is invoked asynchronously by the GUI event loop (well, I don't know if there's a good value for stacklevel in that case, so perhaps we can just live with it...).

hershen commented:
Oh, that's a nice idea in _warn_external.

Sorry, I didn't follow your key_press_handler example. Why would _warn_external not work well in that case?

And a git question - is there a way to continue working on this branch without polluting the history (for example changing warnings.warn(___, stacklevel=2) to _warn_external)? Or is it better to just create a new branch from master?

anntzer commented:
Perhaps key_press_handler is not the best example but I remember looking into doing a global replace and thinking that some of the warn calls would become more confusing if using _warn_external.  I don't have a case right now, someone just needs to do a careful review of the changes.
re: git: you can create a throwaway branch here if you don't want to lose your work, then you can git reset --hard HEAD~ this first branch and rework from there.

anntzer commented:
Sounds good.

anntzer commented on lib/matplotlib/font_manager.py:
```diff
@@ -874,7 +874,7 @@ def json_dump(data, filename):
             json.dump(data, fh, cls=JSONEncoder, indent=2)
         except OSError as e:
-            warnings.warn('Could not save font_manager cache {}'.format(e))
+            warnings.warn('Could not save font_manager cache {}'.format(e), stacklevel=2)
```
For example, I don't think that this warning would get clearer if using _warn_external (or even stacklevel=2).

hershen replied:
OK, I see what you mean.

Is it safe to assume that this warning will be "triggered" when importing pyplot? If so, then does it make sense to print the user line of code that imports pyplot i.e. ```import matplotlib.pyplot as plt```?

I tried to manipulate _warn_external to do this without success. If this solution is acceptable, any ideas how to achieve that?

anntzer replied:
> Is it safe to assume that this warning will be "triggered" when importing pyplot?

Yes

> If so, then does it make sense to print the user line of code that imports pyplot i.e. import matplotlib.pyplot as plt?

I wouldn't overthink it (there isn't anything really smart to do in this case IMO) and would just trigger a warning at this line; or perhaps is actually makes sense to use logging.warning instead of warnings.warn here in fact, now that I think of it...

anntzer replied:
After further thought, I convinced myself that that should exactly be the criterion for `warnings` vs `logging`: by showing the line of offending code, does that point the user to something they should modify?

timhoffm commented:
@hershen still interested to work on this?

hershen commented:
I am. Sorry it got set aside for so long. I'll try to finish it this weekend.

timhoffm commented:
No hurry. Just wanted to make sure this is not orphaned.

hershen commented:
I (finally) switched the `warnings.warn` to either `logging.warning` or `cbook._warn_external`. I followed @anntzer's advice and tried to choose `cbook._warn_external` when I thought the reason for the warning is the user's code or if the warning will help the user find the instance in the code where the warning occurred and `logging.warning` when the user can't fix his code. Please have a look if the choice of function makes sense. Please pay particular attention to the following places where I felt even less sure which function to choose:

- lib/matplotlib/artist.py (lines 404, 855)
- lib/matplotlib/backend_manager.py (263)
- lib/matplotlib/font_manager.py (1230, 1239)
- lib/matplotlib/lines.py (475)
- lib/matplotlib/mathtext.py (all)
- lib/matplotlib/textpath.py (321, 370)
- lib/matplotlib/tight_layout.py (230)
- lib/matplotlib/axes/_base.py (3912)
- lib/matplotlib/backends/backend_pgf.py (402)
- lib/matplotlib/backends/backend_webagg_core.py (243)
- lib/matplotlib/backends/qt_editor/formlayout.py (97, 268)
- lib/matplotlib/style/core.py (193)
- lib/mplt_toolkits/mplot3d/axes3d.py (1081)

anntzer commented on lib/matplotlib/_constrained_layout.py:
```diff
@@ -221,7 +221,7 @@ def do_constrained_layout(fig, renderer, h_pad, w_pad,
                     ax._set_position(newpos, which='original')
         else:
-            warnings.warn('constrained_layout not applied.  At least '
+            cbook._warn_ext('constrained_layout not applied.  At least '
```
typo

anntzer commented on lib/matplotlib/artist.py:
```diff
@@ -850,7 +852,7 @@ def set_rasterized(self, rasterized):
         """
         if rasterized and not hasattr(self.draw, "_supports_rasterization"):
-            warnings.warn("Rasterization of '%s' will be ignored" % self)
+            logging.warning("Rasterization of '%s' will be ignored" % self)
```
looks like cbook._warn_external

anntzer commented on lib/matplotlib/backends/backend_webagg_core.py:
```diff
@@ -241,8 +240,8 @@ def handle_event(self, event):
     def handle_unknown_event(self, event):
-        warnings.warn('Unhandled message type {0}. {1}'.format(
-            event['type'], event), stacklevel=2)
+        cbook._warn_external('Unhandled message type {0}. {1}'.format(
```
I think this should use logging (it's basically an event handler and thus likely to be called asynchronously anyways).

anntzer commented on lib/matplotlib/backends/backend_wx.py:
```diff
@@ -1573,10 +1572,10 @@ def save_figure(self, *args):
                 # looks like they forgot to set the image type drop
                 # down, going with the extension.
-                warnings.warn(
+                cbook._warn_external(
```
logging
(this is most likely called throgh gui interaction)

anntzer commented on lib/matplotlib/backends/backend_wx.py:
```diff
@@ -1837,10 +1836,10 @@ def trigger(self, *args):
             # looks like they forgot to set the image type drop
             # down, going with the extension.
-            warnings.warn(
+            cbook._warn_external(
```
same as above

anntzer commented on lib/matplotlib/backend_bases.py:
```diff
@@ -2396,7 +2395,7 @@ def _get_uniform_gridstate(ticks):
                 ax.set_yscale('log')
             except ValueError as exc:
-                warnings.warn(str(exc))
+                cbook._warn_external(str(exc))
```
logging (most likely called through gui interaction)

anntzer commented on lib/matplotlib/backend_bases.py:
```diff
@@ -2409,7 +2408,7 @@ def _get_uniform_gridstate(ticks):
                 ax.set_xscale('log')
             except ValueError as exc:
-                warnings.warn(str(exc))
+                cbook._warn_external(str(exc))
```
same as above

anntzer commented on lib/matplotlib/lines.py:
```diff
@@ -472,7 +472,7 @@ def contains(self, mouseevent):
         # Convert pick radius from points to pixels
         if self.figure is None:
-            warnings.warn('no figure set when check if mouse is on line')
+            logging.warning('no figure set when check if mouse is on line')
```
I think warn_external here (can be callback or direct call).

hershen replied:
Isn't this gui interaction?

anntzer replied:
I was thinking that one can also call contains() directly, but actually that's almost certainly much rarer, so logging is fine.

anntzer commented on lib/matplotlib/mathtext.py:
```diff
@@ -1595,7 +1590,7 @@ def _set_glue(self, x, sign, totals, error_type):
         if o == 0:
             if len(self.children):
-                warnings.warn(
+                cbook._warn_external(
```
also logging?  a bit borderline but consistent with the rest of the mathtext module

anntzer commented on lib/matplotlib/__init__.py:
```diff
@@ -244,7 +244,7 @@ def _set_logger_verbose_level(level_str='silent', file_str='sys.stdout'):
                 # if this fails, we will just write to stdout
             except IOError:
-                warnings.warn('could not open log file "{0}"'
+                logging.warning('could not open log file "{0}"'
```
throughout the PR:
You should use the module-level logger
`_log = logging.getLogger(__name__)` then `_log.warning(...)`.

hershen replied:
- Ah, right. I even read that I should do that, but for some reason it didn't register.
- Thanks for pointing that out. I wasn't really paying attention to that.

anntzer commented:
Thanks for the thorough changes!  Left some comments.

anntzer commented on lib/matplotlib/mathtext.py:
```diff
@@ -40,10 +40,12 @@
 ##############################################################################
 # FONTS
 
+
```
no empty line here?

hershen replied:
I added it because I got ```E302 expected 2 blank lines, found 1 ``` from flake8 when it wasn't there.
Is there a better solution?

anntzer replied:
Putting the blank line *before* the ##### seems more logical?  Should hopefully also placate the style checker.

hershen replied:
You're right. That fixed it.

anntzer commented on lib/matplotlib/tight_layout.py:
```diff
@@ -172,15 +172,15 @@ def auto_adjust_subplotpars(
-        matplotlib.cbook._warn_external(
-            'The left and right margins cannot be made large enough to '
-            'accommodate all axes decorations. ')
+        matplotlib.cbook._warn_external('The left and right margins cannot '
```
just do `from matplotlib import cbook` at the top?

hershen replied:
Fixed, thanks.

hershen commented:
Thanks so much @anntzer for the dedicated review.

Should I squash all the commits together?

And should I be alarmed that the codecov/patch test is failing?

anntzer commented:
Please do squash.
I wouldn't overly worry about the coverage failure for this specific patch as it's unlikely that we're triggering all these warnings in the test suite and it doesn't seem particularly contructive to try to achieve that.

hershen commented:
Squashed.

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -449,29 +449,71 @@ Then they will receive messages like::
+`warnings.warn`) is that `cbook._warn_external` should be used for things the
+user must change to stop the warning (typically in the source), whereas
+`logging.warning` can be more persistent.  Moreover, note that
+`cbook._warn_extrenal` will by default only emit a given warning *once* for
```
Typo: extrenal

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -449,29 +449,71 @@ Then they will receive messages like::
-layouting or rendering) should only log at this level.
+By default, `warnings.warn` displays the line of code that has the `warn` call.
+This usually isn't more informative than the warning message itself. Therefore,
+Matplotlib uses `cbook._warn_external` which uses `warnings.warn`, but it goes
```
Remove 'it'

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -449,29 +449,71 @@ Then they will receive messages like::
+up the stack and displays the first line of code outside of Matplotlib.
+For example, for the module::
+
+    #in my_matplotlib_module.py
```
Space after `#`

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -449,29 +449,71 @@ Then they will receive messages like::
+running the script::
+
+    from matplotlib import my_matplotlib_module
+    my_matplotlib_module.set_range(0,0) #set range
```
Space after `,` and `#`; two spaces before `#`.

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -449,29 +449,71 @@ Then they will receive messages like::
+    UserWarning: Attempting to set identical bottom==top
+    warnings.warn('Attempting to set identical bottom==top')
+
+Modifiying the module to use `cbook._warn_extermal`::
```
Typo: Modifiying; extermal.

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -449,29 +449,71 @@ Then they will receive messages like::
+and running the same script will display::
+
+  UserWarning: Attempting to set identical bottom==top
+  my_matplotlib_module.set_range(0,0) #set range
```
Same space changes as above.

QuLogic commented on lib/matplotlib/backend_bases.py:
```diff
@@ -38,10 +38,10 @@
 from weakref import WeakKeyDictionary
 
 import numpy as np
+import logging
```
This belongs in the stdlib section above.

QuLogic commented on lib/matplotlib/backend_managers.py:
```diff
@@ -4,13 +4,14 @@
 """
 
-import warnings
-
```
Keep blank line.

QuLogic commented on lib/matplotlib/backends/backend_webagg_core.py:
```diff
@@ -15,15 +15,16 @@
 
 import numpy as np
 import tornado
+import logging
```
Move to above stdlib section of imports.

QuLogic commented on lib/matplotlib/bezier.py:
```diff
@@ -2,7 +2,7 @@
 """
 
-import warnings
+import matplotlib.cbook as cbook
```
Move down to below numpy (can also add a blank line after numpy)

QuLogic commented on lib/matplotlib/colorbar.py:
```diff
@@ -20,7 +20,7 @@
 
 import logging
-import warnings
+import matplotlib.cbook as cbook
```
Move down to matplotlib section below.

QuLogic commented on lib/matplotlib/backends/backend_wx.py:
```diff
@@ -29,6 +28,9 @@
 from matplotlib import cbook, rcParams, backend_tools
 
 import wx
+import logging
```
Move up to stdlib section of imports.

QuLogic commented on lib/matplotlib/dates.py:
```diff
@@ -1311,11 +1311,13 @@ def get_locator(self, dmin, dmax):
+                                     'range. It may be necessary to add an '
+                                     'interval value to the '
+                                     'AutoDateLocator\'s intervald '
+                                     'dictionary. Defaulting to {'
```
Funny place to break; better to keep `{0}` together.

QuLogic commented on lib/matplotlib/image.py:
```diff
@@ -364,9 +363,9 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
-                            warnings.warn(
-                                "Casting input data from '{0}' to 'float64'"
-                                "for imshow".format(A.dtype))
+                            cbook._warn_external("Casting input data from '{"
```
Same here; keep `{0}` together.

hershen commented:
Thanks for the thorough review @QuLogic. I addressed all your comments (and fixed a few other places where I had the same issues).

I'll squash again when you're happy.

hershen commented:
Sorry, but what did I do to make the travis/no language test angry?

QuLogic commented:
Seems to be a bug in Homebrew's NumPy package. Feel free to squash, but it probably won't fix it.

hershen commented:
OK. Squashed. 

The travis issues here seem to be common to all the commits in PRs in the last ~15 hours, so I assume it's not something I'm did wrong.

hershen commented:
OK. Squashed. 

The travis issues here seem to be common to all the commits in PRs in the last ~15 hours, so I assume it's not something I'm did wrong.

hershen commented:
OK. Squashed. 

The travis issues here seem to be common to all the commits in PRs in the last ~15 hours, so I assume it's not something I'm did wrong.

hershen commented:
Rebased.

timhoffm commented on lib/matplotlib/backends/backend_pgf.py:
```diff
@@ -400,9 +399,8 @@ def __init__(self, figure, fh, dummy=False):
-                warnings.warn("streamed pgf-code does not support raster "
-                              "graphics, consider using the pgf-to-pdf option",
-                              UserWarning, stacklevel=2)
+                _log.warning("streamed pgf-code does not support raster "
```
Isn't this of the category "things the that the user must change to stop the warning" and thus should be `_warn_external`?

timhoffm commented on lib/matplotlib/backends/backend_webagg_core.py:
```diff
@@ -241,8 +242,8 @@ def handle_event(self, event):
     def handle_unknown_event(self, event):
-        warnings.warn('Unhandled message type {0}. {1}'.format(
-            event['type'], event), stacklevel=2)
+        _log.warning('Unhandled message type {0}. {1}'.format(
```
also `_warn_external`?

hershen replied:
This was changed to ```_log.warning``` following @anntzer's suggestion.

anntzer replied:
In practice this will almost always get called asynchronously by the webserver event handler so "external" will point to the call to mainloop(), which is not really helpful.

timhoffm commented on lib/matplotlib/backends/backend_wx.py:
```diff
@@ -1573,10 +1575,10 @@ def save_figure(self, *args):
                 # looks like they forgot to set the image type drop
                 # down, going with the extension.
-                warnings.warn(
+                _log.warning(
```
improper use -> `_warn_external`?

hershen replied:
This was changed to ```_log.warning``` following @anntzer's suggestion.

anntzer replied:
In practice this will almost always get called through a wx GUI event handler so "external" will point to the call to mainloop(), which is not really helpful.

timhoffm commented on lib/matplotlib/backends/backend_wx.py:
```diff
@@ -1837,10 +1839,10 @@ def trigger(self, *args):
             # looks like they forgot to set the image type drop
             # down, going with the extension.
-            warnings.warn(
+            _log.warning(
```
improper use -> `_warn_external`?

hershen replied:
This was changed to ```_log.warning``` following @anntzer's suggestion.

anntzer replied:
In practice this will almost always get called through a wx GUI event handler so "external" will point to the call to mainloop(), which is not really helpful.

timhoffm commented on lib/matplotlib/backends/qt_editor/formlayout.py:
```diff
@@ -93,7 +94,7 @@ def to_qcolor(color):
         rgba = mcolors.to_rgba(color)
     except ValueError:
-        warnings.warn('Ignoring invalid color %r' % color, stacklevel=2)
+        _log.warning('Ignoring invalid color %r' % color)
```
improper value passed -> `_warn_external`?

timhoffm commented on lib/matplotlib/figure.py:
```diff
@@ -433,10 +432,10 @@ def show(self, warn=True):
-            warnings.warn('Matplotlib is currently using %s, which is a '
-                          'non-GUI backend, so cannot show the figure.'
-                          % get_backend())
+                    and warn):
```
Why the additional spaces? [PEP-8](https://www.python.org/dev/peps/pep-0008/#indentation) example suggests:

~~~
if (this_is_one_thing
        and that_is_another_thing):
    do_something()
~~~

timhoffm commented on lib/mpl_toolkits/axes_grid1/inset_locator.py:
```diff
@@ -504,9 +503,10 @@ def inset_axes(parent_axes, width, height, loc='upper right',
-                          "Using bbox_to_anchor=(0,0,1,1) now.")
+            cbook._warn_external("Using the axes or figure transform "
+                                 "requires a bounding box in the respective "
+                                 "coordinates. Using bbox_to_anchor=(0,0,1,"
```
I wouldn't break the parenthesis for better readability in the code. Would break befrore "Using" here even though the line will get quite short.

timhoffm commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -1074,7 +1080,7 @@ def mouse_init(self, rotate_btn=1, zoom_btn=3):
             self._cids = [c1, c2, c3]
         else:
-            warnings.warn(
+            _log.warning(
```
Is this not also a case the user should do something about (and thus `_warn_external`).

timhoffm commented on setupext.py:
```diff
@@ -765,7 +767,7 @@ def include_dirs_hook():
         if not has_include_file(
                 ext.include_dirs, os.path.join("numpy", "arrayobject.h")):
-            warnings.warn(
+            _log.warning(
```
User action required -> `_warn_extenal`.

anntzer replied:
This doesn't live in the matplotlib package so "external" will point to this very file; showing the source won't help.

hershen commented:
Thanks @timhoffm for the review!
Some of your suggestions were what I had originally, but @anntzer suggested I change them.
As I'm not familiar enough with the codebase, I'll wait for you two to agree.

anntzer commented:
Agreed with @timhoffm on some, left comments on others.

hershen commented:
Added @timhoffm's suggestions.

jklymak commented:
@anntzer, can you merge if your review still holds?

kbrose commented:
Can the title of this PR be updated to reflect its contents? I had a hard time finding it because its title ("Added stacklevel=2 to all warnings.warn calls (issue 10643)") leaves out the important part of changing every warning.warn call.
